Implementation of design for form page

Sujey-Naik opened the pull request:
A simple UI is built using flutter for form page according to the design given. Used RadioListTile, CheckboxTile, TextFormField and dropdownFormField widgets to built this UI.

Sujey-Naik committed:
Implementation of design for form page

A simple UI is built using flutter for form page according to the design given. Used RadioListTile, CheckboxTile, TextFormField and dropdownFormField widgets to built this UI.

Amulya-Joshi commented on lib/assets/constants.dart:
```diff
@@ -0,0 +1,42 @@
+Color elevatedButtonDisabledColor = Colors.grey.shade300;
+const selectedWidgetColor = Colors.indigoAccent;
+
+class Response {
```
Please add this in a different file.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,399 @@
+  }
+}
+
+enum Gender {
```
Please add this in a different file.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,399 @@
+  OTHERS,
+}
+
+extension GenderExtension on Gender {
```
Create a different file for extensions

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,399 @@
+  ENGLAND,
+}
+
+extension CountryExtension on Country {
```
Same as above.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,399 @@
+  }
+}
+
+enum Country {
```
Same as above.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,399 @@
+  }
+}
+
+Country? _getCountry(String? name) {
```
Add this in a file for helper methods.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/main.dart:
```diff
@@ -1,115 +1,116 @@
-          children: <Widget>[
-            const Text(
-              'You have pushed the button this many times:',
+          materialTapTargetSize: MaterialTapTargetSize.shrinkWrap,
```
Why?

Sujey-Naik replied:
I used radioListTile only as it was working after applying the visual density property. Here, materialTapTargetSize is a property that sets the area around the widget to catch gestures.
When it's value is shrinkWrap, the area of the widget to catch gestures is minimum i.e it shrinks to the size of the widget. 

I did learn more about it now and understood that it really doesn't effect the visible size of the widget. So I tried removing the property and tested, output was correct.

Amulya-Joshi replied:
Discussed further about this offline.

Amulya-Joshi commented on lib/main.dart:
```diff
@@ -1,115 +1,116 @@
-            const Text(
-              'You have pushed the button this many times:',
+          materialTapTargetSize: MaterialTapTargetSize.shrinkWrap,
+          visualDensity: const VisualDensity(
```
Why?

Sujey-Naik replied:
Visual Density represents the compactness(how closely they are placed to each other) of various UI components. It is a unitless value. Its value ranges from -4 to 4, where -4 indicates denser (closer) and 4 indicated least denser ( more expanded). Here minimumDensity indicates the value of -4 for visual density. By default, the visual density value is 0.0. I gave the minimum visual density to keep the UI components as close as possible i.e for radioListTile(title and icon) and checkListTile(title and icon)

Amulya-Joshi commented on lib/main.dart:
```diff
@@ -1,115 +1,116 @@
+              horizontal: VisualDensity.minimumDensity,
+              vertical: VisualDensity.minimumDensity),
+          listTileTheme: const ListTileThemeData(
+            minLeadingWidth: 0.0,
```
Why?

Sujey-Naik replied:
minLeadingWidth is the property that is used to specify the minimum width of the leading icon in the listTile. The higher the value, the width of the icon increases which makes it far away from the title. So I kept it to zero so that the leading icon and title should be as close as possible.

Sujey-Naik committed:
Created different files for enums and helper methods

As mentioned in the PR review separate files for enums and it's extension functions are created. Also separate file to define response object is created. After testing once again, MaterialTapTargetSize property is removed as it was not required.

Amulya-Joshi commented on lib/assets/constants.dart:
```diff
@@ -0,0 +1,27 @@
+const double dividerThickness = 2.0;
+const String appBarTitle = "Form";
+const Color elevatedButtonEnabledColor = Colors.indigoAccent;
+Color elevatedButtonDisabledColor = Colors.grey.shade300;
```
Why is this not a constant?

Amulya-Joshi replied:
Please add color constants in a different class.

Sujey-Naik replied:
> Why is this not a constant?

Colors.grey.shade300 cannot be a constant because it cannot be fully determined at compile-time. The Colors.grey is a named constant, but the shade300 property is not a constant, but rather a getter method that computes a value at runtime based on the current color space.

Sujey-Naik replied:
> Please add color constants in a different class.

Done.

Amulya-Joshi commented on lib/assets/constants.dart:
```diff
@@ -0,0 +1,27 @@
+const double appBarElevation = 0.0;
+const double alertDialogBorderRadius = 8.0;
+const double dividerThickness = 2.0;
+const String appBarTitle = "Form";
```
This does not belong here. We use localization for strings. This will be done in a future PR.

Sujey-Naik replied:
Ok.

Amulya-Joshi commented on lib/assets/constants.dart:
```diff
@@ -0,0 +1,27 @@
+const SizedBox sizedBoxResponsePage = SizedBox(
+  height: 16.0,
+);
+const double appBarElevation = 0.0;
```
This should be added in the theme.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/assets/constants.dart:
```diff
@@ -0,0 +1,27 @@
+const double alertDialogBorderRadius = 8.0;
+const double dividerThickness = 2.0;
+const String appBarTitle = "Form";
+const Color elevatedButtonEnabledColor = Colors.indigoAccent;
```
Is the hex value here exactly as per the design? Please check again. Let me know if you need help with this.

Amulya-Joshi replied:
Colors(#FFFFFFFF);

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/assets/constants.dart:
```diff
@@ -0,0 +1,27 @@
+const String appBarTitle = "Form";
+const Color elevatedButtonEnabledColor = Colors.indigoAccent;
+Color elevatedButtonDisabledColor = Colors.grey.shade300;
+const selectedWidgetColor = Colors.indigoAccent;
```
This should be in the time as well. You can create `primaryColor` in the constants file and use it in the theme.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/assets/enums.dart:
IMO `resources` is a better name for the folder than `assets`. We usually store images and icons under assets.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/assets/extensions.dart:
```diff
@@ -0,0 +1,31 @@
+import 'enums.dart';
```
Create a extension class for this and store this in a folder called `extension`. Structuring your file is a important practice as it makes code more readable.

Amulya-Joshi replied:
Discussed file structuring offline with example.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/assets/extensions.dart:
```diff
@@ -0,0 +1,31 @@
+  }
+}
+
+extension CountryExtension on Country {
```
Why is this an extensions whereas `getCountry` is a helper method?

Amulya-Joshi commented on lib/assets/response_object.dart:
```diff
@@ -0,0 +1,14 @@
+class Response {
```
This is a model. This should be under the model folder.

Sujey-Naik replied:
Done.

Amulya-Joshi commented on lib/assets/response_object.dart:
```diff
@@ -0,0 +1,14 @@
+      required this.dob,
+      required this.gender,
+      required this.opinion,
+      required this.country});
```
Missing `,` for formatting.

Sujey-Naik replied:
Changes Done.

Amulya-Joshi commented on lib/assets/response_object.dart:
```diff
@@ -0,0 +1,14 @@
+class Response {
+  String name;
+  String dob;
+  String? gender;
```
Why is this nullable and required?

Sujey-Naik replied:
This is nullable because the gender value before selecting is null in RadioListTile.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+
+class _HomePageState extends State<HomePage> {
+  final _formKey = GlobalKey<FormState>();
+  final nameController = TextEditingController();
```
Why is this not private?

Sujey-Naik replied:
Changes Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+  final nameController = TextEditingController();
+  final _dobController = TextEditingController();
+  final _opinionController = TextEditingController();
+  bool _acceptTerms = false;
```
This is not the right way to name booleans. `isTermsAccepted` is a better variable name.

Sujey-Naik replied:
Changes Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+  Country? _selectedCountry;
+
+  void _submitForm() {
+    if (_formKey.currentState!.validate()) {
```
```suggestion
    if (_formKey.currentState && !_formKey.currentState.validate()) return;
```

Amulya-Joshi replied:
This is a better approach than using `!` as it will prevent crashes.

Sujey-Naik replied:
Changes Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+
+  @override
+  Widget build(BuildContext context) {
+    var headlineSmall = Theme.of(context).textTheme.headlineSmall;
```
Why var?

Amulya-Joshi replied:
What is the difference between `final`, `var` and `const`?

Sujey-Naik replied:
final and const are used to set a value to a variable only once. Var is used for the variable that changes its value. final is a runtime constant and const variable should be calculated at compile time.

Sujey-Naik replied:
> Why var?

Changed to final.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+  }
+
+  RadioListTile _createRadioListTile(Gender gender) {
+    var bodySmall = Theme.of(context).textTheme.bodySmall;
```
Why var?

Sujey-Naik replied:
Changed to final.

Amulya-Joshi commented on lib/main.dart:
```diff
@@ -1,115 +1,117 @@
   Widget build(BuildContext context) {
     return MaterialApp(
-      title: 'Flutter Demo',
       theme: ThemeData(
```
Please create a separate  class for this theme.

Sujey-Naik replied:
Changes Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+    return RadioListTile(
+      activeColor: selectedWidgetColor,
+      title: Text(
+        gender.name!,
```
Why force null check here?

Sujey-Naik replied:
Changes Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+  }
+
+  AlertDialog _createAlertDialog() {
+    var bodySmall = Theme.of(context).textTheme.bodySmall;
```
Why var. Please check if this style is as per design.

Sujey-Naik replied:
changed to final.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+
+  Column _createTextFormFieldForName() {
+    var bodySmall = Theme.of(context).textTheme.bodySmall;
+    var bodyLarge = Theme.of(context).textTheme.bodyLarge;
```
Why var?

Sujey-Naik replied:
changed to final.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+              lastDate: DateTime(endDate),
+            );
+            if (date != null) {
+              _dobController.text = DateFormat("dd/MM/yyyy").format(date);
```
This can be an extension.

Sujey-Naik replied:
Changes Done.

Amulya-Joshi commented on lib/home_page.dart:
```diff
@@ -0,0 +1,348 @@
+            }
+          },
+          validator: (value) {
+            if (value!.isEmpty) {
```
Why `!`?

Sujey-Naik replied:
Changes Done.

Amulya-Joshi commented on lib/home_page.dart:
This file is too long. Please structure it well.

Sujey-Naik replied:
Changes Done.

Sujey-Naik committed:
Implementation of structural changes for form page

All files are stored in different folders as per their requirements. Code is formatted for better structure.

Sujey-Naik commented on lib/resources/widgets/text_form_field_for_name.dart:
```diff
@@ -0,0 +1,40 @@
+          style: bodySmall,
+          validator: (value) {
+            if (value != null && (value.isEmpty || value.length < IntConstants.nameMinLength)) {
+              return "\u{24D8} Name should have at-least ${IntConstants.nameMinLength} characters";
```
The Unicode for the alert circle(circle with an exclamation mark inside it) was not found. Tried searching for the Unicode on the internet but was not found.

Amulya-Joshi replied:
Can you suggest an alternative? What if we had a use case when we had to show a material icon in the error message?

Sujey-Naik replied:
The alternative is to use a row widget that has an icon with the error text.

Amulya-Joshi replied:
@Sujey-Naik This answer is not acceptable. This lacks context. Please explain in detail.

Sujey-Naik replied:
There is textEditingController for the text form field of name(nameController). addListener function is added to nameController. I did add a validation function as the parameter for addListener(This method invokes the parameter function every time there is a change in the nameController). There is a boolean value called isNameValid which is initially true. If isNameValid is true, then the Row widget with an error icon and error text is not displayed. If it is false, then the row widget is displayed. In the validation function based on criteria for validation, I change the value of isNameValid to true or false. So whenever there is interaction with the text form field of the name, the validation is done and the error message is shown based on isNameValid boolean value. 
When we submit the form, I first check if the text form field is empty(when user has made no interaction with it). If text form field is empty, I change the value of isNameValid to false and call setState() method to show the error message. Then I check whether the isNameValid and isDobValid is true. Only if it is true then the form will be submitted. Else, the form will not be submitted.
So the error message is controlled using a boolean value.

Sujey-Naik commented on lib/resources/widgets/text_form_field_for_dob.dart:
```diff
@@ -0,0 +1,75 @@
+              builder: (context, child) => Theme(
+                data: ThemeData().copyWith(
+                    colorScheme: ColorScheme.light(
+                      primary: AppColors.datePicker.color,
```
The border color of the initial date, it is the same as the primary color. But it should be black in color. I tried changing color values of different parameters of color schema and also tried searching on internet but did not found an answer to it.

Amulya-Joshi replied:
Here is the code for the widget implementation https://github.com/flutter/flutter/blob/20fa32e129bd9b01f60f6152467c6e534eeb5e0d/packages/flutter/lib/src/material/calendar_date_picker.dart#L936. Can you tell me now why you weren't able to solve the issue. Also, can you suggest alternative approach.

Sujey-Naik replied:
The color for the border of the initial date cannot be changed because it takes the color that has been assigned to the primary parameter. This is how it is designed. So, it's color cannot be changed.

Amulya-Joshi replied:
@Sujey-Naik Not accepting this answer. Please explain why this is not possible.

Sujey-Naik replied:
There is a primary parameter for the theme data of the date picker in the color scheme. So this primary parameter takes a color value. The color that is assigned to the primary value is taken by the header part of the calendar and also the date that is selected is of primary color. The border of initial date also uses the same primary color. So, I cannot change it to a different color. When I try to change the color for primary parameter, to make the border color of initial date change, the color assigned to selected date and header of the calendar also changes. So it is not possible to make changes in the border color of initial date.

Sujey-Naik committed:
Changed the error icon used

The error icon was changed for the text form field in the page. The error icon was shown using Icon widget available in flutter.
